False positive `unnecessary-comprehension` when modifying a dict to a list of tuples

### Steps to reproduce

Given a file `a.py`:

```python
def main():
    mydict = {"foo": 1, "bar": 2}
    return [(k, v) for k, v in mydict.items()]
```

Pylint will complain about an unnecessary comprehension, even though the comprehension is modifying the type from a dictionary to a list of tuples.

### Current behavior

Result of `pylint a.py`:
```
$ pylint a.py 
************* Module example
example.py:3:0: R1721: Unnecessary use of a comprehension (unnecessary-comprehension)
```

### Expected behavior

No errors.

### pylint --version output

Result of `pylint --version` output:

```
$ pylint --version output
pylint 2.8.2
astroid 2.5.6
Python 3.8.7 (default, Jan 22 2021, 22:54:27) 
[Clang 12.0.0 (clang-1200.0.32.28)]
```

**EDIT:** I guess alternatively it might be cleaner to use `list(mydict.items())` instead. That feels like it might be a separate issue altogether, and I'm not sure if there aren't other false positives that can't be solved similarly. It's not that the comprehension is unnecessary, it's that there's a nicer way to do it. Maybe a new warning might also be a good idea?

--- a/ChangeLog
+++ b/ChangeLog
@@ -80,6 +80,10 @@ modules are added.
 * Make ``using-constant-test`` detect constant tests consisting of list literals like ``[]`` and
   ``[1, 2, 3]``.
 
+* Improved error message of ``unnecessary-comprehension`` checker by providing code suggestion.
+
+  Closes #4499
+
 
 What's New in Pylint 2.8.2?
 ===========================

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -6,7 +6,7 @@ import copy
 import itertools
 import tokenize
 from functools import reduce
-from typing import List, Union, cast
+from typing import List, Optional, Tuple, Union, cast
 
 import astroid
 
@@ -1393,17 +1393,18 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         else:
             return
         if expr_list == target_list != []:
-            args = None
+            args: Optional[Tuple[str]] = None
+            inferred = utils.safe_infer(node.iter)
             if isinstance(node.parent, astroid.DictComp) and isinstance(
-                utils.safe_infer(node.iter), astroid.objects.DictItems
+                inferred, astroid.objects.DictItems
             ):
                 args = (f"{node.iter.func.expr.as_string()}",)
-            if (
+            elif (
                 isinstance(node.parent, astroid.ListComp)
-                and isinstance(utils.safe_infer(node.iter), astroid.List)
+                and isinstance(inferred, astroid.List)
             ) or (
                 isinstance(node.parent, astroid.SetComp)
-                and isinstance(utils.safe_infer(node.iter), astroid.Set)
+                and isinstance(inferred, astroid.Set)
             ):
                 args = (f"{node.iter.as_string()}",)
             if args:

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -1393,15 +1393,11 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         else:
             return
         if expr_list == target_list != []:
+            args = None
             if isinstance(node.parent, astroid.DictComp) and isinstance(
                 utils.safe_infer(node.iter), astroid.objects.DictItems
             ):
-                self.add_message(
-                    "unnecessary-comprehension",
-                    node=node,
-                    args=(f"{node.iter.func.expr.as_string()}",),
-                )
-                return
+                args = (f"{node.iter.func.expr.as_string()}",)
             if (
                 isinstance(node.parent, astroid.ListComp)
                 and isinstance(utils.safe_infer(node.iter), astroid.List)
@@ -1409,11 +1405,9 @@ class RefactoringChecker(checkers.BaseTokenChecker):
                 isinstance(node.parent, astroid.SetComp)
                 and isinstance(utils.safe_infer(node.iter), astroid.Set)
             ):
-                self.add_message(
-                    "unnecessary-comprehension",
-                    node=node,
-                    args=(f"{node.iter.as_string()}",),
-                )
+                args = (f"{node.iter.as_string()}",)
+            if args:
+                self.add_message("unnecessary-comprehension", node=node, args=args)
                 return
 
             if isinstance(node.parent, astroid.DictComp):

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -284,7 +284,7 @@ class RefactoringChecker(checkers.BaseTokenChecker):
             "raise statement.",
         ),
         "R1721": (
-            "Unnecessary use of a comprehension",
+            "Unnecessary use of a comprehension, use %s instead.",
             "unnecessary-comprehension",
             "Instead of using an identity comprehension, "
             "consider using the list, dict or set constructor. "
@@ -1343,11 +1343,11 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         self._check_consider_using_join(node)
 
     @utils.check_messages("unnecessary-comprehension", "unnecessary-dict-index-lookup")
-    def visit_comprehension(self, node):
+    def visit_comprehension(self, node: astroid.Comprehension) -> None:
         self._check_unnecessary_comprehension(node)
         self._check_unnecessary_dict_index_lookup(node)
 
-    def _check_unnecessary_comprehension(self, node):
+    def _check_unnecessary_comprehension(self, node: astroid.Comprehension) -> None:
         if (
             isinstance(node.parent, astroid.GeneratorExp)
             or len(node.ifs) != 0
@@ -1393,7 +1393,18 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         else:
             return
         if expr_list == target_list != []:
-            self.add_message("unnecessary-comprehension", node=node)
+            if isinstance(node.parent, astroid.DictComp):
+                func = "dict"
+            elif isinstance(node.parent, astroid.ListComp):
+                func = "list"
+            else:
+                func = "set"
+
+            self.add_message(
+                "unnecessary-comprehension",
+                node=node,
+                args=(f"{func}({node.iter.as_string()})",),
+            )
 
     @staticmethod
     def _is_and_or_ternary(node):

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -1393,12 +1393,37 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         else:
             return
         if expr_list == target_list != []:
+            if isinstance(node.parent, astroid.DictComp) and isinstance(
+                utils.safe_infer(node.iter), astroid.objects.DictItems
+            ):
+                self.add_message(
+                    "unnecessary-comprehension",
+                    node=node,
+                    args=(f"{node.iter.func.expr.as_string()}",),
+                )
+                return
+            if (
+                isinstance(node.parent, astroid.ListComp)
+                and isinstance(utils.safe_infer(node.iter), astroid.List)
+            ) or (
+                isinstance(node.parent, astroid.SetComp)
+                and isinstance(utils.safe_infer(node.iter), astroid.Set)
+            ):
+                self.add_message(
+                    "unnecessary-comprehension",
+                    node=node,
+                    args=(f"{node.iter.as_string()}",),
+                )
+                return
+
             if isinstance(node.parent, astroid.DictComp):
                 func = "dict"
             elif isinstance(node.parent, astroid.ListComp):
                 func = "list"
-            else:
+            elif isinstance(node.parent, astroid.SetComp):
                 func = "set"
+            else:
+                return
 
             self.add_message(
                 "unnecessary-comprehension",

--- a/tests/functional/u/unnecessary/unnecessary_comprehension.py
+++ b/tests/functional/u/unnecessary/unnecessary_comprehension.py
@@ -1,4 +1,4 @@
-# pylint: disable=undefined-variable, pointless-statement, missing-docstring, line-too-long
+# pylint: disable=undefined-variable, pointless-statement, missing-docstring, line-too-long, expression-not-assigned
 # For name-reference see https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries
 
 # List comprehensions
@@ -12,6 +12,9 @@
 [y for x in iterable for y in x]  # exclude nested comprehensions
 [2 * x for x in iterable]  # exclude useful comprehensions
 [(x, y, 1) for x, y in iterable]  # exclude useful comprehensions
+# Test case for issue #4499
+a_dict = dict()
+[(k, v) for k, v in a_dict.items()]  # [unnecessary-comprehension]
 
 # Set comprehensions
 {x for x in iterable}  # [unnecessary-comprehension]

--- a/tests/functional/u/unnecessary/unnecessary_comprehension.py
+++ b/tests/functional/u/unnecessary/unnecessary_comprehension.py
@@ -37,3 +37,13 @@ a_dict = dict()
 {y: z for x in iterable for y, z in x}  # exclude nested comprehensions
 {x: 1 for x in iterable}  # expression != target_list
 {2 * x: 3 + x for x in iterable}  # exclude useful comprehensions
+
+# Some additional tests on helptext -- when object is already a list/set/dict
+my_list = list()
+my_dict = dict()
+my_set = set()
+
+[elem for elem in my_list]  # [unnecessary-comprehension]
+{k: v for k, v in my_dict.items()} # [unnecessary-comprehension]
+{k: my_dict[k] for k in my_dict} # [consider-using-dict-items]
+{elem for elem in my_set}  # [unnecessary-comprehension]

--- a/tests/functional/u/unnecessary/unnecessary_comprehension.txt
+++ b/tests/functional/u/unnecessary/unnecessary_comprehension.txt
@@ -7,3 +7,7 @@ unnecessary-comprehension:23:0::Unnecessary use of a comprehension, use set(iter
 unnecessary-comprehension:24:0::Unnecessary use of a comprehension, use set(iterable) instead.
 unnecessary-comprehension:32:0::Unnecessary use of a comprehension, use dict(iterable) instead.
 unnecessary-comprehension:34:0::Unnecessary use of a comprehension, use dict(iterable) instead.
+unnecessary-comprehension:46:0::Unnecessary use of a comprehension, use my_list instead.
+unnecessary-comprehension:47:0::Unnecessary use of a comprehension, use my_dict instead.
+consider-using-dict-items:48:0::Consider iterating with .items()
+unnecessary-comprehension:49:0::Unnecessary use of a comprehension, use my_set instead.

--- a/tests/functional/u/unnecessary/unnecessary_comprehension.txt
+++ b/tests/functional/u/unnecessary/unnecessary_comprehension.txt
@@ -1,8 +1,9 @@
-unnecessary-comprehension:5:0::Unnecessary use of a comprehension
-unnecessary-comprehension:8:0::Unnecessary use of a comprehension
-unnecessary-comprehension:9:0::Unnecessary use of a comprehension
-unnecessary-comprehension:17:0::Unnecessary use of a comprehension
-unnecessary-comprehension:20:0::Unnecessary use of a comprehension
-unnecessary-comprehension:21:0::Unnecessary use of a comprehension
-unnecessary-comprehension:29:0::Unnecessary use of a comprehension
-unnecessary-comprehension:31:0::Unnecessary use of a comprehension
+unnecessary-comprehension:5:0::Unnecessary use of a comprehension, use list(iterable) instead.
+unnecessary-comprehension:8:0::Unnecessary use of a comprehension, use list(iterable) instead.
+unnecessary-comprehension:9:0::Unnecessary use of a comprehension, use list(iterable) instead.
+unnecessary-comprehension:17:0::Unnecessary use of a comprehension, use list(a_dict.items()) instead.
+unnecessary-comprehension:20:0::Unnecessary use of a comprehension, use set(iterable) instead.
+unnecessary-comprehension:23:0::Unnecessary use of a comprehension, use set(iterable) instead.
+unnecessary-comprehension:24:0::Unnecessary use of a comprehension, use set(iterable) instead.
+unnecessary-comprehension:32:0::Unnecessary use of a comprehension, use dict(iterable) instead.
+unnecessary-comprehension:34:0::Unnecessary use of a comprehension, use dict(iterable) instead.

